contracts/bridges/BaseBridgeReceiver.sol

```diff
@@ -0,0 +1,134 @@
+    event ProposalExecuted(uint id);
+
+    address public mainnetTimelock;
+    address public l2Timelock;
+    bool public initialized;
+
+    uint public proposalCount;
+
+    struct Proposal {
+        uint id;
+        address[] targets;
+        uint[] values;
+        string[] signatures;
+        bytes[] calldatas;
+        uint eta;
+        bool executed;
+    }
+
+    mapping (uint => Proposal) public proposals;
+
+    enum ProposalState {
+        Queued,
+        Expired,
+        Executed
+    }
+
+    function initialize(address _mainnetTimelock, address _l2Timelock) external {
+        if (initialized) revert AlreadyInitialized();
+        mainnetTimelock = _mainnetTimelock;
+        l2Timelock = _l2Timelock;
+        initialized = true;
+        emit Initialized(_mainnetTimelock, _l2Timelock);
+    }
+
+    function acceptL2TimelockAdmin() external {
+        if (msg.sender != l2Timelock) revert Unauthorized();
+        ITimelock(l2Timelock).acceptAdmin();
+    }
+
+    function setL2Timelock(address newTimelock) public {
+        if (msg.sender != l2Timelock) revert Unauthorized();
+        address oldL2Timelock = l2Timelock;
+        l2Timelock = newTimelock;
+        emit NewL2Timelock(oldL2Timelock, newTimelock);
+    }
+
+    function setMainnetTimelock(address newTimelock) public {
+        if (msg.sender != l2Timelock) revert Unauthorized();
+        address oldMainnetTimelock = mainnetTimelock;
+        mainnetTimelock = newTimelock;
+        emit NewMainnetTimelock(oldMainnetTimelock, newTimelock);
+    }
+
+    function processMessage(
+        address messageSender,
+        bytes calldata data
+    ) internal {
+        if (messageSender != mainnetTimelock) revert Unauthorized();
+
+        address[] memory targets;
+        uint256[] memory values;
+        string[] memory signatures;
+        bytes[] memory calldatas;
+
+        (targets, values, signatures, calldatas) = abi.decode(
+            data,
+            (address[], uint256[], string[], bytes[])
+        );
+
+        if (values.length != targets.length) revert BadData();
+        if (signatures.length != targets.length) revert BadData();
+        if (calldatas.length != targets.length) revert BadData();
+
+        uint delay = ITimelock(l2Timelock).delay();
+        uint eta = block.timestamp + delay;
+
+        for (uint8 i = 0; i < targets.length; ) {
+            ITimelock(l2Timelock).queueTransaction(targets[i], values[i], signatures[i], calldatas[i], eta);
+            unchecked { i++; }
+        }
+
+        proposalCount++;
+        Proposal memory proposal = Proposal({
+            id: proposalCount,
+            targets: targets,
+            values: values,
+            signatures: signatures,
+            calldatas: calldatas,
+            eta: eta,
+            executed: false
+        });
+
+        proposals[proposal.id] = proposal;
+        emit ProposalCreated(messageSender, proposal.id, targets, values, signatures, calldatas, eta);
```

Comment:
Or alternatively, the second argument of the event can just be the `Proposal` struct.